pageserver/src/tenant/storage_layer/image_layer.rs

```diff
@@ -436,8 +439,12 @@ impl ImageLayerInner {
                 )
                 .await
                 .with_context(|| format!("failed to read value from offset {}", offset))?;
-            let value = Bytes::from(blob);
-
+            let value = if is_rel_data_key(key) && blob.len() < BLCKSZ as usize {
+                let decompressed = lz4_flex::block::decompress(&blob, BLCKSZ as usize)?;
+                Bytes::from(decompressed)
+            } else {
+                Bytes::from(blob)
+            };
```

Comment:
To distinguish compressed and uncompressed mages, let's add an explicit tag byte for that instead of this hack with the size. And bump up the `format_version` number.

In order to be able to roll back the release if something goes wrong, let's deploy this in two stages:
1. Add support for reading the new format
2. Once 1. has been deployed for a while, start writing the new format

Reply:
> With or without compression, I would like to use the "image-with-a-hole" format throughout the system. 

Please notice that only regular pages may have hole. SLRU pages doesn't hold them.
Frankly speaking I do not think that supporting punched images is so good idea.
It significantly complicates code and gives almost nothing if pages are compressed.

Reply:
>To distinguish compressed and uncompressed mages, let's add an explicit tag byte for that instead of this hack with the size. And bump up the format_version number.


I agree with this proposal. It gives us better safety and flexibility.
But there is one stupid problem, for which I do not know good solution: how to store this tag.
This is how page is stored in image layer now:
```
        if is_rel_data_key(key) {
            let compressed = lz4_flex::block::compress(img);
            if compressed.len() < img.len() {
                off = self.blob_writer.write_blob(&compressed).await?;
            } else {
                off = self.blob_writer.write_blob(img).await?;
            }
        } else {
            off = self.blob_writer.write_blob(img).await?;
        }
```
So we just call `write_blob` which appears  sequence of bytes to the file.
Now assume that we want also to store some tag (1 byte).
There are two possibilities:
1. Add `blob_writer.write_byte()` method to separately write this tag
2. Prepend it to the stored image (blob). But it requires allocation and copying of ~8kb. May be not so expensive. comparing with compression overhead... But looks ugly.

Are there some better alternatives?

Reply:
@hlinnaka  do you think that we need to add this extra byte for each page image or just one byte in summary?
It seems to be no sense to support different compression algorithms for different pages.
Certainly we in any case need to distinguish compressed and not compressed pages. But what's wrong will locking at BLOB size? Wee can enforce that compression is performed only for BLCKSZ page images and that compressed version always has smaller size than original (otherwise we just do not store compressed image).

Reply:
Looking at the size is complicated and error prone. You have to be careful that the compressed size never happens to be equal to BLCKSZ, for example.

I'd suggest adding a variant of `write_blob` that writes a one byte followed by the buffer. Or an array of buffers, like `writev(2)` does, and hope that the compiler inlines it into something fast.

Yet another approach would be to compress *all* blobs in the file. Then you also don't need a header byte. I think a header byte is better, though.

Reply:
> Looking at the size is complicated and error prone. You have to be careful that the compressed size never happens to be equal to BLCKSZ, for example.

Hmmm...
Not sure that I am convinced by this argument. Comparing size is not more complex than any other comparison, for example comparison of format version. Concerning corner cases, i.e. BLCKSZ - we want compression to make some profit, right? So if it returns exactly the same size then why do we need to store compressed image?

So from my point of view this check is quite obvious and safe.

I already has impelled solution with storing compression algorithm in layer metadata (summary).
I can now try two complement alternative solution with storing extra byte for each BLOB and check if there is some overhead. I do not expect to get some noticeable overhead. But from my point of view it is much more complicated  solution than just looking size given us no  advantages (extra flexibility, better compression,...).

Reply:
> I'd suggest adding a variant of write_blob that writes a one byte followed by the buffer. Or an array of buffers, like writev(2) does, and hope that the compiler inlines it into something fast.

There are less problems with write than with read. In case of write I can call write_all for this tag. 
But BlobCursor has no such method. May be move decompression to blob_io? In this case there is no need to add extra methods...

Reply:
Ok, I have implemented approach with header byte